mmtk/src/active_plan.rs

```diff
@@ -71,29 +84,23 @@ impl ActivePlan<JuliaVM> for VMActivePlan {
     }
 }
 
-#[no_mangle]
-pub extern "C" fn get_next_mutator_tls() -> OpaquePointer {
-    let mutators = MUTATORS.read().unwrap();
-
-    let mutator_idx = unsafe { get_next_julia_mutator() };
-    let mutator = mutators.get(mutator_idx);
-
-    let res = match mutator {
-        Some(m) => {
-            let mutator = unsafe { get_mutator_from_ref(*m) };
+// Expose the mutator iterator so they can be used in C.
 
-            unsafe { (*mutator).mutator_tls.0 .0 }
-        }
-        None => {
-            unsafe { reset_mutator_count() }
-            OpaquePointer::from_address(unsafe { Address::zero() })
-        }
-    };
+#[no_mangle]
+pub extern "C" fn new_mutator_iterator() -> Box<JuliaMutatorIterator<'static>> {
+    let guard = MUTATORS.read().unwrap();
+    Box::new(JuliaMutatorIterator::new(guard))
+}
```

Comment:
```suggestion
pub extern "C" fn new_mutator_iterator() -> *mut JuliaMutatorIterator<'static> {
    let guard = MUTATORS.read().unwrap();
    Box::into_raw(Box::new(JuliaMutatorIterator::new(guard)))
}
```

The layout of `Box<T>` is not transparent (otherwise it should have the `#[repr(transparent)]` annotation), so we cannot assume it has the same layout as `void*`.  In fact, `Box` is defined as:

```rs
pub struct Box<
    T: ?Sized,
    A: Allocator = Global,
>(Unique<T>, A);
```

The safe way is using `Box::into_raw` to convert it to a raw pointer so that it can be given to C code.

Reply:
Done